### PR TITLE
Feature/Better component init

### LINKED ISSUE #28: Long startup time
I configured one of my areas very basically:

```
kuche:
    control_lights: true
    control_climate: false
    control_media: false
    automatic_lights:
      entities:
        - light.kuche_spule
```

The `binary_sensor.area_kuche` turns to on after walking in front of the motion sensor. 
image

But the light doesn't turn on. This happens on a other area as well. The logs don't provide any related errors.

Edit:

Added
`custom_components.magic_areas: debug`
to logger. Maybe I get more information about whats going on.

Edit 2:

What does this mean?
`2020-10-23 20:14:58 INFO (SyncWorker_11) [custom_components.magic_areas.binary_sensor] Küche is on passive mode.`

### PR DESCRIPTION
**Pull request by jseidl**
Implementing proper component initialization using `RestoreEntity` and `async_added_to_hass` to prevent Area state change upon restart of Home Assistant.

This removes entirely the need for `passive_start` which was a workaround for the lack of restore.

Closes #28

### CODE REVIEW

**Comment by jseidl** [comment]
Removed `passive_start` from config and implemented restore ability on the `AreaPresenceBinarySensor`. Still have to implement on the others.

**Comment by sti0** [comment]
Got this error on startup:
`2020-10-26 22:30:35 ERROR (MainThread) [homeassistant.components.binary_sensor] Error while setting up magic_areas platform for binary_sensor
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 201, in _async_setup_platform
    await asyncio.gather(*pending)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 310, in async_add_entities
    await asyncio.gather(*tasks)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 481, in _async_add_entity
    await entity.add_to_platform_finish()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 521, in add_to_platform_finish
    await self.async_added_to_hass()
  File "/config/custom_components/magic_areas/binary_sensor.py", line 256, in async_added_to_hass
    _LOGGER.debug(f"Area restored: {self.slug}")
AttributeError: 'AreaPresenceBinarySensor' object has no attribute 'slug'`

**Comment by jseidl** [comment]
That should be `self.area.slug` Gonna fix that in a bit.

File: custom_components/magic_areas/binary_sensor.py [hunk 1]
```diff
@@ -253,11 +237,63 @@ def device_class(self):
...
+        else:
+            self.hass.bus.async_listen_once(
+                EVENT_HOMEASSISTANT_STARTED, self._setup_listeners
+            )
+
+        last_state = await self.async_get_last_state()
+        is_new_entry = last_state is None  # newly added to HA
+
+        if is_new_entry:
+            _LOGGER.debug(f"New area detected: {self.slug}")
```
Comment (sti0) [inline custom_components/magic_areas/binary_sensor.py:253]:
```suggestion
            _LOGGER.debug(f"New area detected: {self.area.slug}")

```

File: custom_components/magic_areas/binary_sensor.py [hunk 2]
```diff
@@ -253,11 +237,63 @@ def device_class(self):
...
+            self._update_state()
+        else:
+            _LOGGER.debug(f"Area restored: {self.slug}")
```
Comment (sti0) [inline custom_components/magic_areas/binary_sensor.py:256]:
```suggestion
            _LOGGER.debug(f"Area restored: {self.area.slug}")

```

**Comment by sti0** [comment]
Bugreport:
Area is occupied and the lights are on. The `disable_entity` is off. After switching the `disable_entity` to on, the lights stay on. After switching the `disable_entity` back to off, the lights turn off.

**Comment by jseidl** [comment]
If you turn on lights in a non-occupied area, eventually `update_interval` will kick in, see that the area is not occupied and proceed on turning the lights off. `disable_entity` and `sleep_entity` state change also trigger an area state check.

Disable to off turning lights off is weird, unless the room was already clear. Will investigate

**Comment by jseidl** [comment]
Added `RestoreEntity` capabilities to both the Presence `binary_sensor` and the Presence Hold `switch`. The aggregators don't look like would benefit from state restore as running the averaging / grouping on start is not detrimental (at least looks like it so far).

Added hook to `EVENT_HOMEASSISTANT_STARTED` to setup all the state trackers.

Tested on my test install and seems to be working so far. Nothing on my logs and working as intended. Gonna deploy on my home install and test for a few days before merging since a lot of code was moved around on this one.

**Comment by jseidl** [comment]
Been running this branch a couple of days. Seems to be stable. Will merge into `develop` branch this evening

**Comment by sti0** [comment]
> Been running this branch a couple of days. Seems to be stable. Will merge into `develop` branch this evening

I've been testing this branch today as well, but I still facing the issue mentioned in [my comment](https://github.com/jseidl/hass-magic_areas/pull/43#issuecomment-717442117) a couple of days ago.

After some testing I found the issue.
I configured a `disable_state: "off"` and it doesn't work. If I remove this configuration and it works with default state = on, just fine. So there is a bug with a custom `disable_state`.

File: custom_components/magic_areas/binary_sensor.py [hunk 3]
```diff
@@ -253,24 +237,83 @@ def device_class(self):
...
+        self.tracking_listeners.extend([remove_presence, remove_interval])
+
+    def _remove_listeners(self):
+        while self.tracking_listeners:
+            remove_listener = self.tracking_listeners.pop()
+            remove_listener()
+
+    def autolight_disable_state_change(self, entity_id, from_state, to_state):
+
         if to_state.state == STATE_OFF:
```
Comment (sti0) [inline custom_components/magic_areas/binary_sensor.py:298]: doesn't respect the custom `disable_state`

Comment (jseidl) [reply custom_components/magic_areas/binary_sensor.py:298]: Good catch! Will push a fix soon

Comment (jseidl) [reply custom_components/magic_areas/binary_sensor.py:298]: Actually this bug is on the master so should be a `hotfix/`. Will merge this one and push a hotfix to the master.

**Comment by KTibow** [comment]
You can use draft PRs, BTW, instead of putting WIP in the title and using a bot. Draft PRs can't be merged until marked as ready for review.

**Pull request merged**